### PR TITLE
Drop Node 8 and Node 10 support.

### PR DESCRIPTION
**Pull request by bengl**
### What does this PR do?
* Drops Node.js v8 and v10 support.
* Gets rid of all the code I could find that was Node 8 or 10 specific.
* Updates some dependencies.
* Increases minimum version support ranges for plugins (since those plugins added support for Node 12 later on.
* Stops testing on Node 8 and 10 in CI.
* Lint fixes due to updated version of `eslint`.
* Since Node 12 makes it hard to require package.json files, added the ability to do it manually, and added that everywhere that was previously doing `require('${name}/package.json')`.

### Motivation

Both are EOL. It's time.

### Additional Notes
I tried to make each individual change into its own commit so that it can be reviewed separately. Please have a look commit-by-commit.

### CODE REVIEW

File: packages/dd-trace/src/require-package-json.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
+'use strict'
+
+const path = require('path')
+const fs = require('fs')
+
+/**
+ * Given a package name and a module to start from, find a package's
+ * package.json file, parses it, and returns the result.
+ *
+ * Equivalent to require(`${name}/package.json`) prior to Node 12.
```
Comment (rochdev) [inline packages/dd-trace/src/require-package-json.js:10]: Since we already support newer versions of Node, why is this change needed for this PR? Is it a bug fix for something that should already be working and isn't? If that's the case, should it be its own PR? It's not a large change but since it touches the loader I'd move it out if it's not related to dropping Node 8/10.

File: benchmark/stubs/span.js [hunk 2]
```diff
@@ -34,7 +34,7 @@ const span = {
     _traceFlags: {},
     _name: 'operation'
   }),
-  _startTime: 1500000000000.123456,
+  _startTime: 1500000000.123456,
```
Comment (rochdev) [inline benchmark/stubs/span.js:37]: Why this change?

File: .eslintrc.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
   "parserOptions": {
-    "ecmaVersion": 2018
+    "ecmaVersion": 2021
```
Comment (rochdev) [inline .eslintrc.json:3]: Node 12.0.0 was released before that. Should we use an older version to make sure we don't use constructs that may not be supported?

File: .eslintrc.json [hunk 4]
```diff
@@ -14,6 +14,14 @@
     "no-var": 2,
     "no-console": 2,
     "prefer-const": 2,
-    "object-curly-spacing": [2, "always"]
+    "object-curly-spacing": [2, "always"],
```
Comment (rochdev) [inline .eslintrc.json:17]: Let's move any new rules to a different PRs since I wouldn't want to block just because of these (I don't agree with some of them).

**Review by rochdev — CHANGES_REQUESTED** [review]

File: scripts/junit_report.js [hunk 5]
```diff
@@ -1,13 +1,8 @@
 /* eslint-disable no-console */
 
-const semver = require('semver')
 const { execSync } = require('child_process')
 
 function uploadJUnitXMLReport () {
-  if (semver.lt(process.version, '10.24.1')) {
```
[the remark this answers is not in the document]
Comment (juan-fernandez) [reply scripts/junit_report.js:7]: I think removing it is fine.

File: .circleci/config.yml [hunk 6]
```diff
@@ -917,11 +883,12 @@ jobs:
         environment:
           - PLUGINS=next
 
+# does oracle actually need node 10?
```
[the remark this answers is not in the document]
Comment (bengl) [reply .circleci/config.yml:886]: It looks like I was. This comment can be deleted. Will do.

File: packages/datadog-plugin-http/src/client.js [hunk 7]
```diff
@@ -338,13 +337,11 @@ module.exports = [
...
       patch.call(this, http, 'request', tracer, config)
-      if (semver.satisfies(process.version, '>=8')) {
-        /**
-         * In newer Node versions references internal to modules, such as `http(s).get` calling `http(s).request`, do
-         * not use externally patched versions, which is why we need to also patch `get` here separately.
-         */
-        patch.call(this, http, 'get', tracer, config)
-      }
+      /**
+       * In newer Node versions references internal to modules, such as `http(s).get` calling `http(s).request`, do
```
[the remark this answers is not in the document]
Comment (bengl) [reply packages/datadog-plugin-http/src/client.js:341]: Sure.

File: packages/dd-trace/src/require-package-json.js (hunk 1 above)
Comment (bengl) [reply packages/dd-trace/src/require-package-json.js:10]: >  Is it a bug fix for something that should already be working and isn't? If that's the case, should it be its own PR?

Sort of. Because we generally test things on our minimum support versions, we weren't testing on Node 12 for a lot of things, which introduces the new behaviour with the `"exports"` field. This means that yes, we had a bug here that was in our testing blindspot. By moving to Node 12, it's no longer in our testing blind spot, and we need a fix for CI to pass at all, so here we are.

I could do this in a separate PR but:

1. We'd have to test on Node 12 in that PR. Without having a full testing matrix, that's a bit difficult right now.
2. That PR would have to land before this one.

File: benchmark/stubs/span.js (hunk 2 above)
Comment (bengl) [reply benchmark/stubs/span.js:37]: newer eslint breaks on numbers that would be truncated when parsed.

File: .eslintrc.json (hunk 4 above)
Comment (bengl) [reply .eslintrc.json:17]: Sure, that's fine.

File: .circleci/config.yml [hunk 8]
```diff
@@ -190,7 +190,7 @@ node-upstream-base: &node-upstream-base
 
 prebuild-linux-base: &prebuild-linux-base
   docker:
-    - image: node:10.0.0
+    - image: node:12.0.0
```
[the remark this answers is not in the document]
Comment (bengl) [reply .circleci/config.yml:193]: That's ... unexpected. I'll change it back and add a comment saying something about that.

File: .eslintrc.json (hunk 3 above)
Comment (bengl) [reply .eslintrc.json:3]: Some ES2020 things are supported by Node 12, because supported is often added before the versions finalize. This setting simply allows the parser to operate correctly on these features. If we want to disallow some, those should be in rules, not the parser. Anyway, tests should tell us if we're using unsupported features.

File: scripts/install_plugin_modules.js [hunk 9]
```diff
@@ -99,9 +101,13 @@ function assertPackage (name, version, dependency, external) {
 function assertIndex (name, version) {
   const index = `'use strict'
 
+const fs = require('fs')
```
Comment (rochdev) [inline scripts/install_plugin_modules.js:104]: Why are `fs` and `path` needed here?

Comment (bengl) [reply scripts/install_plugin_modules.js:104]: Previously, `requirePackageJson` was inlined. I'll fix that now.

**Review by rochdev — APPROVED** [review]

**Review by rochdev — APPROVED** [review]
